### ISSUE TITLE
The shared library of C++ API lacks of operation symbols on windows.

### ISSUE THREAD

**Issue by EternalSaga** [tensorflow/tensorflow#30246]
**System information**
- Windows 10 1809:
- TensorFlow installed from (source or binary): Source
- TensorFlow version: 1.14.0
- Python version: 3.7.3
- Bazel version (if compiling from source):  0.25.2
- GCC/Compiler version (if compiling from source): visual studio 2017

**Describe the problem**
I tried to build the [label image c++ demo](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/label_image/main.cc) and I got lots of linking error which cannot find the tf operation symbols. Such as:
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2001	unresolved external symbol "public: __cdecl tensorflow::ops::Subtract::Subtract(class tensorflow::Scope const &,class tensorflow::Input,class tensorflow::Input)" (??0Subtract@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@1@Z)	tf1.14.0	D:\ProgrammingAndStudy\C++_proj\tf1.14.0\tf1.14.0\main.obj	1	

**Provide the exact sequence of commands / steps that you executed before running into the problem**
I build the tensorflow 1.14.0 C++ API through the MSYS2 shell after some configeration. 
Although I configured using vs2019 here, but it seems useless. Bazel still choosing the vs2017.
export MSYS_NO_PATHCONV=1
export MSYS2_ARG_CONV_EXCL="*" 
export PATH="/C/Users/yueji/AppData/Local/Programs/Python/Python37":$PATH
alias python='winpty python.exe'
export PATH="/C/machine learning dlls":$PATH
set BAZEL_VS="C:\Program Files (x86)\Microsoft Visual Studio\2019\Community"
set BAZEL_VC="C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC"

And then, I compiled the tf1.14.0 through
bazel build --config=opt //tensorflow:tensorflow_cc
bazel build --config=opt //tensorflow:tensorflow_framework

Finally, I got two libs and two dlls. They are located in C:\Users\user_name\_bazel_yueji\dvw6l3y3\execroot\org_tensorflow\bazel-out\x64_windows-opt\bin\tensorflow
tensorflow.dll.if.lib, tensorflow_framework.dll.if.lib
tensorflow_cc.dll, tensorflow_framework.dll

I configured my vs projects to make sure the headers are properly included, and linked these two libs. 
However there are lots of lnk2001 errors are reported.

**Any other info / logs**
It's a part of error file.

Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2001	unresolved external symbol "public: __cdecl tensorflow::ops::Subtract::Subtract(class tensorflow::Scope const &,class tensorflow::Input,class tensorflow::Input)" (??0Subtract@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@1@Z)	tf1.14.0	D:\ProgrammingAndStudy\C++_proj\tf1.14.0\tf1.14.0\main.obj	1	
Error	LNK2001	unresolved external symbol "public: __cdecl tensorflow::ops::Squeeze::Squeeze(class tensorflow::Scope const &,class tensorflow::Input)" (??0Squeeze@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@@Z)	tf1.14.0	D:\ProgrammingAndStudy\C++_proj\tf1.14.0\tf1.14.0\main.obj	1	
Error	LNK2001	unresolved external symbol "public: __cdecl tensorflow::ops::DecodeGif::DecodeGif(class tensorflow::Scope const &,class tensorflow::Input)" (??0DecodeGif@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@@Z)	tf1.14.0	D:\ProgrammingAndStudy\C++_proj\tf1.14.0\tf1.14.0\main.obj	1	
Error	LNK2001	unresolved external symbol "public: __cdecl tensorflow::ops::TopK::TopK(class tensorflow::Scope const &,class tensorflow::Input,class tensorflow::Input)" (??0TopK@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@1@Z)	tf1.14.0	D:\ProgrammingAndStudy\C++_proj\tf1.14.0\tf1.14.0\main.obj	1

**Comment by EternalSaga** [comment]
I tried tu use dumpbin tools in msvc commandline tools in powershell:
dumpbin.exe /headers .\tensorflow_cc.dll.if.lib | Select-String "ResizeBilinear"
dumpbin.exe /headers .\tensorflow_cc.dll.if.lib | Select-String "Squeeze"
dumpbin.exe /headers .\tensorflow_cc.dll.if.lib | Select-String "DecodeGif"
dumpbin.exe /headers .\tensorflow_cc.dll.if.lib | Select-String "ExpandDims"
Unfortunately, nothing was returned.

I also tried to build the library directly: bazel build --config=opt //tensorflow:tensorflow_cc.lib
The result is the same.

**Comment by gadagashwini** [comment]
@EternalSaga looks like similar issue [#24885](https://github.com/tensorflow/tensorflow/issues/24885). Please let us know if that helps you. Thanks!

**Comment by EternalSaga** [comment]
@gadagashwini  Thank you for your advice. But it seems useless for me. I cannot find any solutions about my problem.

**Comment by acezzz** [comment]
please try to see if  https://github.com/guikarist/tensorflow-windows-build-script.git could help.

**Comment by jeffhwang02** [comment]
Same issue here. I tried tf_exported_symbol_msvc.ids method and def_fp.write method too. but both failed to solve my case.  Current my builting tensorflow_cc.dll has only c-api, there are no c++ APIs.

**Comment by jeffhwang02** [comment]
I hope Bazel developmemt team solve this issue by comparing cmake, it has worked well in making c++ api dll for windows.

**Comment by meteorcloudy** [comment]
So the reason you don't have all the ops symbols is because the generated source code of ops do not include the `__declspec(export)` descriptor for classes like `DecodeGif` on Windows. Check `bazel-bin/tensorflow/cc/ops/image_ops.h`

Both CMake and Bazel can parse objects to get global symbols and use it to export symbols in DLL, we use this method to export symbols for tensorflow_cc.dll. But unfortunately, that doesn't work for symbols in C++ Class.

I tried to solve this by adding `clean_dep("//tensorflow/core:ops"),` at 
https://github.com/tensorflow/tensorflow/blob/ba5dddcf708e1762b924e4cd7ef07c54b34708f5/tensorflow/tensorflow.bzl#L1812
But it didn't work because the missing symbols are class methods.

I really don't have any idea to solve this for the shared library of C++ API except adding the `__declspec(export)` descriptor  while generating the source code of ops.

**Comment by meteorcloudy** [comment]
Wait, I think I should have added `clean_dep("//tensorflow/cc:cc_ops")` instead of `clean_dep("//tensorflow/core:ops")`. Let me try again

**Comment by meteorcloudy** [comment]
I think I have found a solution.

@EternalSaga @jeffhwang02 
Can you cherry-pick https://github.com/meteorcloudy/tensorflow/commit/4b8e7bef11040c465453e973eb0c6ad43cfd3c96 and rebuild the `//tensorflow:tensorflow_cc.dll` and `//tensorflow:tensorflow_cc.lib`.

With the change in the commit, I can see the symbols in the import library:
```
dumpbin.exe  /headers ./bazel-bin/tensorflow/tensorflow_cc.lib | grep "ExpandDims"
  Symbol name  : ??0ExpandDims@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@1@Z (public: __cdecl tensorflow::ops::ExpandDims::ExpandDims(class tensorflow::Scope const &,class tensorflow::Input,class tensorflow::Input))
  Name         : ??0ExpandDims@ops@tensorflow@@QEAA@AEBVScope@2@VInput@2@1@Z
```

**Comment by meteorcloudy** [comment]
Ping @EternalSaga @jeffhwang02
Can you confirm if my change fixed the symbol issue on Windows?
If so we can merge that change so that more users can benefit from it.

**Comment by meteorcloudy** [comment]
@jeffhwang02 Do you have a github branch of your source code?

**Comment by jeffhwang02** [comment]
> @jeffhwang02 Do you have a github branch of your source code?

Yes  I did it in the same path which I built the tensorflow_cc.dll file well.

**Comment by jeffhwang02** [comment]
> @jeffhwang02 Do you have a github branch of your source code?

@meteorcloudy 
Yes I did it in the same path which I built the tensorflow_cc.dll file well.

**Comment by meteorcloudy** [comment]
@jeffhwang02 
Oh I mean, can you show me the actual code so that I can see what went wrong?
Also which version of Bazel were you using?

**Comment by philipeccles** [comment]
I would just like to say great work @meteorcloudy 

I could not get my application to link because 

"ExpandDims::ExpandsDims was an unresolved external symbol"

I had already used TF_EXPORT on a bunch of other symbols but could not do same thing here ExpandDims::ExpandDims is part of an auto generated file.

I used the changes to BUILD file in tensorflow sub folder (not the top level one, the one with folders "c" "cc" "compiler"). from last comment 6th JUly 2020.

Worked straight out of the box.

TensorFlow 2.3.0. Bazel 3.1.0 VS2019. VS2017 does not work.

Too many issues to mention trying to get TEnsorFlow to build.

But information in this thread is very useful.

**Comment by SuryanarayanaY** [comment]
@EternalSaga ,

At present the support for 1.x versions not active. Could you please confirm whether this is an issue with latest 2.X versions(>=2.7 version).

If its still an issue with latest 2.X versions please let us know with the build command. 

You can find the latest build instructions for Windows at [here](https://www.tensorflow.org/install/source_windows). Thanks!
